### PR TITLE
[ci] Fix GeoCoq after https://github.com/GeoCoq/GeoCoq/issues/12

### PR DESCRIPTION
**Pull request by ejgallego**
We build a bit of math-comp so it works, also patch in the old-style
until the upstream issue is fixed.

Let's see how this works.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
What's with this crazy sed hack, can't we just `make algebra` or something?

**Comment by Zimmi48** [comment]
I would indeed suppose that math-comp authors have support for this kind of things.

**Comment by ejgallego** [comment]
The hack is not so crazy IMO.

But indeed, we could use the opam infrastructure I guess; I guess againt that I originally did like that because I wanted full control over the general `Make` file.

**Comment by ejgallego** [comment]
Anyways the hack has been there for ages so I'd rather not address it in this PR which is meant to be a "quick" fix.

**Comment by ejgallego** [comment]
Also note that the fix should work even after the authors fix the upstream version.

**Comment by ejgallego** [comment]
Something strange is that their Travis suceeded for `master`, maybe Jason packages are outdated?

**Comment by Zimmi48** [comment]
Or maybe, they don't test everything.

**Comment by ejgallego** [comment]
Yeah their running time is suspiciously short.

**Comment by Zimmi48** [comment]
Looks like the sed hack is not working:

```
make[1]: Entering directory '/builds/coq/coq/_build_ci/math-comp/mathcomp'
COQC ssreflect/ssrnotations.v
COQC ssreflect/ssreflect.v
ssreflect/ssrnotations (real: 0.14, user: 0.08, sys: 0.05, mem: 104152 ko)
make[2]: *** No rule to make target 'fingroup/fingroup.vo', needed by 'algebra/ssrnum.vo'.  Stop.
make[2]: *** Waiting for unfinished jobs....
COQC ssreflect/ssrbool.v
ssreflect/ssreflect (real: 0.18, user: 0.14, sys: 0.04, mem: 119436 ko)
ssreflect/ssrbool (real: 0.19, user: 0.12, sys: 0.06, mem: 127068 ko)
Makefile.coq:328: recipe for target 'all' failed
make[1]: *** [all] Error 2
```

**Comment by SkySkimmer** [comment]
Still not working. The other sed is failing now.

**Comment by gares** [comment]
> What's with this crazy sed hack, can't we just make algebra or something?

yYou can `make -C algebra` but you have to `make -C ssreflect && make -C fingroup` first.

File: dev/ci/ci-common.sh [hunk 1]
```diff
@@ -107,3 +107,24 @@ install_ssreflect()
...
 }
+
+# this installs just the ssreflect + algebra library of math-comp
+install_ssralg()
+{
+  echo 'Installing ssreflect' && echo -en 'travis_fold:start:ssr.install\\r'
+
+  git_checkout "${mathcomp_CI_BRANCH}" "${mathcomp_CI_GITURL}" "${mathcomp_CI_DIR}"
+
+  ( cd "${mathcomp_CI_DIR}/mathcomp" && \
```
Comment (gares) [inline dev/ci/ci-common.sh]: no please

**Review by gares — DISMISSED** [review]

File: dev/ci/ci-common.sh (hunk 1 above)
Comment (ejgallego) [reply dev/ci/ci-common.sh]: This hack was already in place and thus orthogonal to the purpose of this PR. Please let's move the discussion to its own issue  / PR.

Comment (ejgallego) [reply dev/ci/ci-common.sh]: This hack was already in place and thus orthogonal to the purpose of this PR. Please let's move the discussion to its own issue  / PR.

Comment (gares) [reply dev/ci/ci-common.sh]: This is new code. Fix *this* pr please. The fix is trivial. I m sorry I did not suggest this before.

File: dev/ci/ci-geocoq.sh [hunk 2]
```diff
@@ -7,4 +7,7 @@ GeoCoq_CI_DIR="${CI_BUILD_DIR}/GeoCoq"
 
 git_checkout "${GeoCoq_CI_BRANCH}" "${GeoCoq_CI_GITURL}" "${GeoCoq_CI_DIR}"
 
+install_ssralg
+
+sed -i.bak "1iSet Nested Proofs Allowed\n" "${GeoCoq_CI_DIR}/Meta_theory/Models/hilbert_to_tarski.v"
```
Comment (Zimmi48) [inline dev/ci/ci-geocoq.sh]: This is missing a dot and is thus making the build fail:

```
Warning: There is no option Nested Proofs Allowed Require Import Setoid.
[unknown-option,option]
File "./Meta_theory/Models/hilbert_to_tarski.v", line 231, characters 0-58:
Error:
Nested proofs are not allowed unless you turn option Nested Proofs Allowed on.
```

**Review by Zimmi48 — COMMENTED** [review]
Given that this PR needs to be fixed once more, and given that an author of math-comp told you how to replace the horrible sed hack with a cleaner way, I suggest you do follow the recommendation while fixing the PR.

**Comment by Zimmi48** [comment]
The nested lemma has been removed so this part of the fix is not needed anymore. However, I don't understand why you have still not taken into account Enrico's proposal to replace the sed hack.

**Comment by ejgallego** [comment]
> I don't understand why you have still not taken into account Enrico's proposal to replace the sed hack.

I already explained it; the `sed` hack has been in place for a long time and it is orthogonal to this PR and not trivial to eliminate; please open a bug if you want it replaced.

**Comment by Zimmi48** [comment]
I don't understand. We are talking about the sed hack that you are introducing, not the one that was already present. And Enrico provided you a replacement. If this replacement doesn't actually work, I'm fine with the current state of the PR. But if it does work, then why not use it? I suppose you either think that we are asking you to also remove the preexisting sed hack or you see why Enrico's drop in replacement won't be enough but you didn't explain why.

**Comment by ejgallego** [comment]
Just a matter of lack of time @Zimmi48

**Comment by Zimmi48** [comment]
Passed here: https://gitlab.com/Zimmi48/coq/-/jobs/73927070

Sorry @ejgallego I had to reintroduce the `Set Nested Proofs Allowed` option I had asked you to remove. Turns out that, contrarily to what was claimed in GeoCoq/GeoCoq#12, the nested lemma has not been removed (cc. GeoCoq/GeoCoq@e140b35f2a18446e8164b7b9a126aba147c38bb9).

**Comment by Zimmi48** [comment]
Wait a minute.

**Comment by Zimmi48** [comment]
So 83bfdcf can be dropped again after all: the nested lemma had not been removed in GeoCoq/GeoCoq@a3ebcab73d65a0baa6f889baa8ec0ec26dc795ca but it has in GeoCoq/GeoCoq@e140b35f2a18446e8164b7b9a126aba147c38bb9. However, this commit didn't remove the corresponding proof so GeoCoq must be broken again right now.

**Comment by ejgallego** [comment]
Yeah we are having trouble with this PR XD

**Comment by Zimmi48** [comment]
Contrary to what I was thinking, GeoCoq is fixed and this PR is finally ready to merge @SkySkimmer (commits squashed).

**Review by SkySkimmer — APPROVED** [review]

**Pull request merged**